sdks/go/pkg/beam/core/runtime/exec/fn.go

```diff
@@ -39,23 +40,47 @@ type MainInput struct {
 	RTracker sdf.RTracker
 }
 
+type bundleFinalizationCallback struct {
+	callback   func() error
+	validUntil time.Time
+}
+
+// bundleFinalizer holds all the user defined callbacks to be run on bundle finalization.
+// Implements typex.BundleFinalization
+type bundleFinalizer struct {
+	callbacks         []bundleFinalizationCallback
+	lastValidCallback time.Time // Used to track when we can safely gc the bundleFinalizer
+}
+
+// RegisterCallback is used to register callbacks during DoFn execution.
+func (bf *bundleFinalizer) RegisterCallback(t time.Duration, cb func() error) {
+	callback := bundleFinalizationCallback{
+		callback:   cb,
+		validUntil: time.Now().Add(t),
+	}
+	bf.callbacks = append(bf.callbacks, callback)
+	if bf.lastValidCallback.Before(callback.validUntil) {
+		bf.lastValidCallback = callback.validUntil
+	}
+}
+
 // Invoke invokes the fn with the given values. The extra values must match the non-main
 // side input and emitters. It returns the direct output, if any.
-func Invoke(ctx context.Context, pn typex.PaneInfo, ws []typex.Window, ts typex.EventTime, fn *funcx.Fn, opt *MainInput, extra ...interface{}) (*FullValue, error) {
+func Invoke(ctx context.Context, pn typex.PaneInfo, ws []typex.Window, ts typex.EventTime, fn *funcx.Fn, opt *MainInput, bf *bundleFinalizer, extra ...interface{}) (*FullValue, error) {
```

Comment:
But what about adding the optional bundleFinalizer on a field on MainInput, like we do the optional RTracker?  (To be clear, they aren't exactly the same notionally, but they're very similar, I can be convinced either way).

Don't forget to update the Design Doc with the implementation attempts and rationale about why we are or are not going with that approach.

Reply:
I would argue against it in this instance. Functionally, the bundleFinalizer is much closer to being an emitter than it is an input - its a tool that is used to produce an output (callbacks) and thus I'd be pretty surprised to find it in the MainInput. FWIW, I'm also kinda surprised to find RTracker in MainInput, though I think there's a significantly stronger case that it should be included since it does modify how we're able to access the input (if I understand its purpose correctly 🤷)

Reply:
I'm in agreement that this is further from being a MainInput than RTracker was. In that case, the intention wasn't just that RTracker modified how the input is accessed, but also that it is the interface for accessing the restriction, which is considered to _be part of_ the input elements. The same way that attaching a key to a value turns them into a single MainInput.

Like for a practical example, say we have an SDF that takes filename strings as elements and byte ranges as the restriction. If we had inputs "foo.txt" [0, 10] and "foo.txt" [10, 20], those are considered separate input elements despite the string portion being the same because they represent different byte ranges due to the restriction. That's why I considered the RTracker (or more specifically, the restriction itself) to be part of the MainInput.